src/js/environments.ts

```diff
@@ -1,71 +1,92 @@
-
-/** @private */
-export const IN_NODE_COMMONJS =
-  IN_NODE &&
-  typeof module !== "undefined" &&
-  typeof module.exports !== "undefined" &&
-  typeof require !== "undefined" &&
-  typeof __dirname !== "undefined";
-
-/** @private */
-export const IN_NODE_ESM = IN_NODE && !IN_NODE_COMMONJS;
-
-/** @private */
-export const IN_BUN = typeof globalThis.Bun !== "undefined";
+/**
+ * Internal runtime environment interface for type safety
+ * @private
+ */
+interface RuntimeEnv {
+  IN_NODE: boolean;
+  IN_NODE_COMMONJS: boolean;
+  IN_NODE_ESM: boolean;
+  IN_BUN: boolean;
+  IN_DENO: boolean;
+  IN_BROWSER: boolean;
+  IN_BROWSER_MAIN_THREAD: boolean;
+  IN_BROWSER_WEB_WORKER: boolean;
+  IN_SAFARI: boolean;
+  IN_SHELL: boolean;
+}
 
-/** @private */
-export const IN_DENO = typeof Deno !== "undefined"; // just in case...
+/**
+ * Internal runtime environment state
+ * @private
+ */
+function getGlobalRuntimeEnv(): RuntimeEnv {
+  if (!globalThis.__PYODIDE_RUNTIME_ENV__) {
+    // Derived flags are computed during initialization
+    const env: RuntimeEnv = {
+      IN_NODE:
+        typeof process === "object" &&
+        typeof process.versions === "object" &&
+        typeof process.versions.node === "string" &&
+        !(process as any).browser,
 
-/** @private */
-export const IN_BROWSER = !IN_NODE && !IN_DENO;
+      IN_NODE_COMMONJS: false, // Derived from IN_NODE
+      IN_NODE_ESM: false, // Derived from IN_NODE
 
-/** @private */
-export const IN_BROWSER_MAIN_THREAD =
-  IN_BROWSER &&
-  typeof window === "object" &&
-  typeof document === "object" &&
-  typeof document.createElement === "function" &&
-  "sessionStorage" in window &&
-  typeof importScripts !== "function";
+      IN_BUN: typeof (globalThis as any).Bun !== "undefined",
+      IN_DENO: typeof (globalThis as any).Deno !== "undefined",
+      IN_BROWSER: true, // Default true, will be updated in derived flags
+      IN_BROWSER_MAIN_THREAD: false, // Derived from IN_BROWSER
+      IN_BROWSER_WEB_WORKER: false, // Derived from IN_BROWSER,
 
-/** @private */
-export const IN_BROWSER_WEB_WORKER =
-  IN_BROWSER && typeof importScripts === "function" && typeof self === "object";
+      IN_SAFARI:
+        typeof navigator === "object" &&
+        typeof (navigator as any).userAgent === "string" &&
+        (navigator as any).userAgent.indexOf("Chrome") == -1 &&
+        (navigator as any).userAgent.indexOf("Safari") > -1,
 
-/** @private */
-export const IN_SAFARI =
-  typeof navigator === "object" &&
-  typeof navigator.userAgent === "string" &&
-  navigator.userAgent.indexOf("Chrome") == -1 &&
-  navigator.userAgent.indexOf("Safari") > -1;
+      IN_SHELL:
+        typeof (globalThis as any).read == "function" &&
+        typeof (globalThis as any).load === "function",
+    };
+    // Update derived flags using the shared function
+    updateDerivedFlags(env);
+    globalThis.__PYODIDE_RUNTIME_ENV__ = env;
```

Comment:
@ryanking13 We shouldn't be adding any global variables. I opened #5917 to remove it.

Reply:
Good point. Adding this to globalThis is not needed.